crates/codegen/parser/runtime/src/query_model.rs

```diff
@@ -0,0 +1,144 @@
+use std::rc::Rc;
+
+use super::kinds::{FieldName, RuleKind, TokenKind};
+
+#[derive(Clone)]
+pub enum Query {
+    Binding(Rc<BindingMatch>),
+    Node(Rc<NodeMatch>),
+    Optional(Rc<OptionalMatch>),
+    Alternatives(Rc<AlternativesMatch>),
+    Sequence(Rc<SequenceMatch>),
+    OneOrMore(Rc<OneOrMoreMatch>),
+    Ellipsis,
+}
+
+impl Query {
+    pub fn parse(text: &str) -> Result<Self, String> {
+        super::query_parser::parse_query(text)
+    }
+
+    pub fn unparse(&self) -> String {
+        match self {
+            Self::Binding(binding) => {
+                format!("@{} {}", binding.name, binding.child.unparse())
+            }
+            Self::Node(node) => {
+                if let Some(child) = &node.child {
+                    format!("[{} {}]", node.id.unparse(), child.unparse())
+                } else {
+                    format!("[{}]", node.id.unparse())
+                }
+            }
+            Self::Optional(optional) => {
+                format!("({})?", optional.child.unparse())
+            }
+            Self::Alternatives(alternatives) => {
+                format!(
+                    "({})",
+                    alternatives
+                        .children
+                        .iter()
+                        .map(|c| c.unparse())
+                        .collect::<Vec<_>>()
+                        .join(" | ")
+                )
+            }
+            Self::Sequence(sequence) => sequence
+                .children
+                .iter()
+                .map(|c| c.unparse())
+                .collect::<Vec<_>>()
+                .join(" "),
+            Self::OneOrMore(one_or_more) => {
+                format!("({})+", one_or_more.child.unparse())
+            }
+            Self::Ellipsis => "...".to_string(),
+        }
+    }
+}
+
+#[derive(Copy, Clone, PartialEq, Eq)]
+pub enum Kind {
```

Comment:
instead of exposing `query_model::Kind` here, as we might need that type in the future for similar reasons, what do you think about exposing it publicly as `kinds::NodeKind` instead?
This is similar to the `kinds.NodeKind` TypeScript type we already export.

Reply:
YAGNI? I don't think we should hoist this in anticipation of maybe wanting it. I will look at limiting the visibility

Reply:
Not sure if YAGNI applies here?
I mean, if we are doing the work regardless of defining and maintaining that type, I suggest exposing it through the `cst` module instead (for organizational/above reasons).
If we can limit its visibility/there is no need to expose it, that also sounds good to me.